### DISCUSSION TITLE
Any plans for a Kernel Density Estimator?

### DISCUSSION THREAD

**Discussion by ndalchau** [mathnet/mathnet-numerics#802] [Ideas]
The title says it all really. Is anyone considering adding KDE functionality?

**Comment by cdrnet** [comment]
No concrete plans so far from my side, but I'd certainly accept a pull request.

**Comment by waynenilsen** [comment]
I'm not sure where this should go but perhaps some variation of below? Got the reference from wikipedia https://en.wikipedia.org/wiki/Kernel_density_estimation

``` c#
public static double KDE(double at, IList<double> sample, double bandwidth)
{
    return sample.Select(x_i => Normal.PDF(0, bandwidth, at - x_i)).Sum() / sample.Count();
}
```

Seems to be doing its job
![image](https://cloud.githubusercontent.com/assets/3612857/12117076/34c14110-b38c-11e5-8b6c-941ae937fcb8.png)

**Comment by Bawanglong** [comment]
I want this too.
I think it would be nice if one could specify the used kernel, with the Gaussian as a default. Automatic bandwidth selection would also be great (and maybe support for d-dimensional samples).

I'll try to come up with an API and make a proposal for that.

**Comment by Bawanglong** [comment]
If somebody wants to discuss an API, I opened a PR, see above.

In terms of functionality: So far, it supports 1D samples. What's still missing is some input checking (e.g. samples null or empty), corner-case testing and automatic bandwidth selection.

**Comment by cdrnet** [comment]
Thanks all for the proposal and first implementation.

I've already pulled in the proposal, but refactored it a bit towards a simpler static and stateless API:
* `KernelDensity.Estimate(double x, double bandwidth, IList<double> samples, Func<double, double> kernel)`
* `KernelDensity.EstimateGaussian(double x, double bandwidth, IList<double> samples)`
* ...

I have not included any functionality yet regarding bandwidth rules, but it seems we could support it nicely with overloads in the future.

**Comment by cdrnet** [comment]
An option would be to add `*Func` variants returning a lambda on x, i.e. `Func<double,double> KernelDensity.EstimateGaussian(double bandwidth, IList<double> samples)`. We already have this pattern e.g. in `Statistics.QuantileFunc`.
